### PR TITLE
Process-wide channel cache for gRPC+aio

### PR DESCRIPTION
**Pull request by justheuristic**
* [x] process-wide grpc channel/stub caching
    * implemented hivemind.utils.grpc.ChannelCache
    * extracted TimedStorage from hivemind.dht.storage to hivmind.utils.timed_storage
    * [x] use in hivemind/client/expert.py
        * [x] verify no performance degradation
    * [x] use in hivemind/dht/protocol.py
        * [x] verify no performance degradation (upd: got faster)
    * [x] use in  hivemind.client.allreduce.py
        * [x] verify no performance degradation
* [ ] dht.get_endpoint() to discover this peer's current public endpoint from peers
* [ ] implement a basic synchronization check for get_dht_time()

### process-wide channel caching
Currently, each time we query a remote peer with gRPC, we have to create a new channel.
In contrast, gRPC best practices [recommend](https://docs.microsoft.com/en-us/aspnet/core/grpc/performance?view=aspnetcore-5.0) reusing channels for multiple rpc calls

hivemind/client/expert.py introduced channel caching, but it had a side-effect of keeping channels open forever

In this PR we implement a process-wide ChannelCache object that keeps track of open channels. The code is largely inspired by https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_simple_stubs.py , but with the added support for grpc.aio channels.

### CODE REVIEW

**Comment by justheuristic** [comment]
upd - from discussion with @mryab  - 
image

we decided to postpone these two to a #98 as they are not essential for ICML experiments

File: hivemind/client/allreduce.py [hunk 1]
```diff
@@ -150,9 +150,7 @@ async def accumulate(self, source: Endpoint, part: torch.Tensor) -> torch.Tensor
         return await self.averaged_part
 
     def _get(self, peer: Endpoint) -> averaging_pb2_grpc.DecentralizedAveragingStub:
```
Comment (mryab) [inline hivemind/client/allreduce.py:152]: I believe _get is slightly overloaded here, how about _get_peer_stub?

File: hivemind/utils/grpc.py [hunk 2]
```diff
@@ -1,17 +1,151 @@
 """
 Utilities for running GRPC services: compile protobuf, patch legacy versions, etc
 """
+from __future__ import annotations
+import os
+import threading
+from typing import NamedTuple, Sequence, Tuple, Optional, Union, Any, Dict, TypeVar, Type
```
Comment (mryab) [inline hivemind/utils/grpc.py:7]: If we're using `from __future__ import annotations`, do we need to import tuple and dict from typing?

File: hivemind/utils/grpc.py [hunk 3]
```diff
@@ -1,17 +1,151 @@
...
+
+            return stubs[stub_type]
+
+    @classmethod
+    def _create_channel(cls, target: Endpoint, aio: bool, options: Sequence[Tuple[str, Any], ...],
+                        channel_credentials: Optional[grpc.ChannelCredentials],
+                        compression: Optional[grpc.Compression]) -> Union[grpc.Channel, grpc.aio.Channel]:
+        namespace = grpc.aio if aio else grpc
+        if channel_credentials is None:
+            logger.debug(f"Creating insecure {namespace} channel with options '{options}' " +
```
Comment (mryab) [inline hivemind/utils/grpc.py:107]:
```suggestion
            logger.debug(f"Creating insecure {namespace} channel with options '{options}' "
```

File: hivemind/utils/grpc.py [hunk 4]
```diff
@@ -1,17 +1,151 @@
...
+                         f"and compression '{compression}'")
+            return namespace.insecure_channel(target, options=options, compression=compression)
+        else:
+            logger.debug(f"Creating secure {namespace} channel with credentials '{channel_credentials}', "
+                         + f"options '{options}' and compression '{compression}'")
```
Comment (mryab) [inline hivemind/utils/grpc.py:112]:
```suggestion
                         f"options '{options}' and compression '{compression}'")
```

File: hivemind/utils/grpc.py [hunk 5]
```diff
@@ -1,17 +1,151 @@
...
+class ChannelCache(TimedStorage[ChannelInfo, Tuple[Union[grpc.Channel, grpc.aio.Channel], Dict]]):
+    """
+    A process-wide cache of gRPC channels, supports both normal and aio channels, secure/insecure channels, etc
+    Based on grpcio internal channel cache by Richard Belleville and Lidi Zheng (thanks!)
+    Unlike TimedStorage, ChannelCache actively evicts stale channels even if the cache is not accessed
+    Unlike grpc._simple_stubs.ChannelCache, this implementation supports aio and does not forcibly close active channels
+    """
+    MAXIMUM_CHANNELS = os.environ.get("GRPC_PYTHON_MANAGED_CHANNEL_MAXIMUM", 4096)
+    EVICTION_PERIOD_SECONDS = os.environ.get("GRPC_PYTHON_MANAGED_CHANNEL_EVICTION_SECONDS", 10 * 60)
+    logger.debug(f"Eviction period = {EVICTION_PERIOD_SECONDS}s, max channels = {MAXIMUM_CHANNELS}")
```
Comment (mryab) [inline hivemind/utils/grpc.py:41]: These constants should be module-level

File: hivemind/utils/grpc.py [hunk 6]
```diff
@@ -1,17 +1,151 @@
...
+
+    _singleton: Optional[ChannelCache] = None
+    _singleton_pid: int = os.getpid()
+    _lock: threading.RLock = threading.RLock()
+    _update_eviction_evt: threading.Event = threading.Event()
+    _eviction_thread: threading.Thread
+    _nearest_expiration_time: DHTExpiration
+    _is_active: bool
```
Comment (mryab) [inline hivemind/utils/grpc.py:49]: These are instance attributes and are declared during init

**Review by mryab — CHANGES_REQUESTED** [review]

File: hivemind/client/allreduce.py (hunk 1 above)
Comment (justheuristic) [reply hivemind/client/allreduce.py:152]: its mostly a convention with DHTProtocol._get, should we rename both?

File: hivemind/utils/grpc.py (hunk 2 above)
Comment (justheuristic) [reply hivemind/utils/grpc.py:7]: i agree, but i would request to keep it as is AND write a todo for #98 , ok?

File: hivemind/utils/grpc.py (hunk 6 above)
Comment (justheuristic) [reply hivemind/utils/grpc.py]: (fixed)

File: hivemind/utils/grpc.py (hunk 5 above)
Comment (justheuristic) [reply hivemind/utils/grpc.py]: (fixed, thanks!)

Comment (justheuristic) [reply hivemind/utils/grpc.py:41]: upd - it complicates testing, added to #98

**Review by mryab — APPROVED** [review]

**Pull request merged**